### PR TITLE
ad9656: add reference project for the ad9656 eval board

### PR DESCRIPTION
**Pull request by hotoleanudan**
Added new reference project for ad9656 eval board coupled with the zcu102 carrier.

Signed-off-by: Dan Hotoleanu <dan.hotoleanu@analog.com>

### CODE REVIEW

File: projects/ad9656_fmc/src/app/ad9656_fmc.c [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+	ad9508_spi_read(dev, 0x0C, &reg_data);
+	if (reg_data != 0x05)
+		return FAILURE;
+	ad9508_spi_read(dev, 0x0D, &reg_data);
+	if (reg_data != 0x00)
+		return FAILURE;
+	
+	// configure 9508 to pass the 125MHz input clock unmodified, so divider = 1
+	// no phase offset
+	ad9508_spi_write(dev, 0x1B, 0x00); // divide ratio[7:0]
```
Comment (amiclaus) [inline projects/ad9656_fmc/src/app/ad9656_fmc.c:231]: Would be nice to use macro defines for the Reg addresses.

Comment (amiclaus) [reply projects/ad9656_fmc/src/app/ad9656_fmc.c:231]: This statement is valid for the entire commit.

File: projects/ad9656_fmc/src/app/app_config.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef APP_CONFIG_H_
+#define APP_CONFIG_H_
+
```
Comment (amiclaus) [inline projects/ad9656_fmc/src/app/app_config.h:42]: This file will be populated in the future when support for new platforms / features will be added?
Otherwise it is redundant.

File: projects/ad9656_fmc/src/devices/adi_hal/parameters.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef _PARAMETERS_H_
+#define _PARAMETERS_H_
+
+#include "app_config.h"
```
Comment (amiclaus) [inline projects/ad9656_fmc/src/devices/adi_hal/parameters.h:43]: You don't need to include `app_config.h`

**Comment by amiclaus** [comment]
Fix please code style issues detected by the travis build.

File: projects/ad9656_fmc/src/app/ad9656_fmc.c [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+#include "delay.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+#define DMA_BUFFER		 0
+
+struct ad9508_init_param {
+	/* SPI */
```
Comment (amiclaus) [inline projects/ad9656_fmc/src/app/ad9656_fmc.c:63]: Just an opinion. I would separate the device driver implementations(ad9553, ad9508) into separate files.
Even though the implementations are minimal, it will be easier to expand them afterwards and add to the `drivers` folder.

**Comment by hotoleanudan** [comment]
Can you please give me a feedback on the last update that I made to this project? Or if there are no more comments on it can I merge it on master branch?

File: projects/ad9656_fmc/src/app/ad9656_fmc.c (hunk 1 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9656_fmc.c]: Changed hardcoded register numbers to defined constants.

File: projects/ad9656_fmc/src/app/ad9656_fmc.c (hunk 4 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9656_fmc.c]: Moved the driver implementation for the ad9553 and ad9508 devices into separate files in ad9656_fmc/src/app folder.

File: projects/ad9656_fmc/src/devices/adi_hal/parameters.h (hunk 3 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/devices/adi_hal/parameters.h]: I would keep this file in order the preserve the file structure of the folder.

File: projects/ad9656_fmc/src/app/app_config.h (hunk 2 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/app_config.h:42]: I discussed with Adrian Costina and he told me that in the near future there will be no support for new platforms / features. But still, I would keep this file in order the preserve the file structure.

File: projects/ad9656_fmc/src/devices/adi_hal/parameters.h (hunk 3 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/devices/adi_hal/parameters.h]: Removed the include.

**Comment by hotoleanudan** [comment]
Changelog v2:
- moved the driver code for ad9508 device from ad9656_fmc/src/app/ad9656_fmc.c to ad9656_fmc/src/app/ad9508.c and ad9656_fmc/src/app/ad9508.h
- moved the driver code for ad9553 device from ad9656_fmc/src/app/ad9656_fmc.c to ad9656_fmc/src/app/ad9553.c and ad9656_fmc/src/app/ad9553.h
- used defined constants for addresses of registers instead of their adresses (names instead of numbers)
- deleted the "include app_config.h" instruction from the parameters.h

**Comment by hotoleanudan** [comment]
Changelog v3:
- Modifed the project files such that it will comply with the AStyle

**Comment by hotoleanudan** [comment]
Changelog v4:

- modified the text indentation for Makefile and src.mk files

File: projects/ad9656_fmc/src/app/ad9508.c [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+#include "delay.h"
+
+/**
+ * @brief Reads from the ad9508 that is contected to the SPI
+ * @param dev - The ad9508 device handler
+ * @param reg_addr - The address of the internal register of the ad9508 chip
+ * @param reg_data - The value read from the internal register
+ * @return SUCCESS if the value was successfully read, FAILURE otherwise
+ */
+int32_t ad9508_spi_read(struct ad9508_dev *dev,
```
Comment (adrimbarean) [inline projects/ad9656_fmc/src/app/ad9508.c:57]: I would rename these `ad9508_reg_read` and `ad9508_reg_write` to be consistent with other drivers. Also, I think these files can be added as drivers in the `drivers` folder and included from there.

File: projects/ad9656_fmc/src/app/ad9508.c [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+	return spi_write_and_read(dev->spi_desc, buf, 3);
+}
+
+/**
+ * @brief Setup the working parameters of the ad9508 chip
+ * @param device - The device handler of the ad9508 chip
+ * @param init_param - Values for the working parameters of ad9508
+ * @return SUCCESS if device is ready for use, FAILURE otherwise
+ */
+int ad9508_setup(struct ad9508_dev **device,
```
Comment (adrimbarean) [inline projects/ad9656_fmc/src/app/ad9508.c:112]: nitpick: 
```suggestion
int32_t ad9508_setup(struct ad9508_dev **device,
```

File: projects/ad9656_fmc/src/app/ad9656_fmc.c [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	else
+		printf("The ad9508 chip successfully configured\n");
+
+	if (ad9553_setup(&ad9553_device, &ad9553_param) != SUCCESS)
+		printf("The ad9553 chip could not be setup!\n");
+	else
+		printf("The ad9553 chip successfully configured\n");
+
+	// ADC
+	ad9656_setup(&ad9656_device, &ad9656_param);
```
Comment (adrimbarean) [inline projects/ad9656_fmc/src/app/ad9656_fmc.c:191]: I would check for this return value as well.

**Comment by hotoleanudan** [comment]
Changelog v5:

- Squased the commits on the branch on the first commit and modified the resulting commit's comment.

File: projects/ad9656_fmc/src/app/ad9508.c (hunk 5 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9508.c:57]: Done in v6

File: projects/ad9656_fmc/src/app/ad9508.c (hunk 6 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9508.c:112]: Done in v6

File: projects/ad9656_fmc/src/app/ad9656_fmc.c (hunk 7 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9656_fmc.c:191]: Done in v6

**Comment by hotoleanudan** [comment]
Changelog v6:

- Changed function names ad9508_spi_read, ad9508_spi_write, ad9553_spi_read and ad9553_spi_write to ad9508_reg_read, ad9508_reg_write, ad9553_reg_read and ad9553_reg_write. Also changed all their occurrences.
- Changed the return type of ad9508_setup and ad9553_setup from int to int32_t.
- Checked the return value of the ad9656_setup function call in ad9656_fmc.c file.

File: projects/ad9656_fmc/src/app/ad9508.c [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+ * @param reg_addr - The address of the internal register of the ad9508 chip
+ * @param reg_data - The value read from the internal register
+ * @return SUCCESS if the value was successfully read, FAILURE otherwise
+ */
+int32_t ad9508_reg_read(struct ad9508_dev *dev,
+			uint16_t reg_addr,
+			uint8_t *reg_data)
+{
+	uint8_t buf[3];
+
```
Comment (mirceacaprioru) [inline projects/ad9656_fmc/src/app/ad9508.c:62]: nitpick: remove extra line

File: projects/ad9656_fmc/src/app/ad9508.c [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+	int32_t ret;
+
+	// the MSB of byte 0 indicates a r/w operation, following 7 bits are the
```
Comment (mirceacaprioru) [inline projects/ad9656_fmc/src/app/ad9508.c:65]: multi line comment
```
/*
 *
 */

File: projects/ad9656_fmc/src/app/ad9508.c [hunk 10]
```diff
@@ -0,0 +1,191 @@
+/***************************************************************************//**
+ * @file ad9508.c
+ * @brief Implementation of AD9508 Driver.
+ * @author DHotolea (dan.hotoleanu@analog.com)
+ ********************************************************************************
+ * Copyright 2020(c) Analog Devices, Inc.
+ *
+ * All rights reserved.
```
Comment (mirceacaprioru) [inline projects/ad9656_fmc/src/app/ad9508.c:8]: looking better here aren't these the driver files ? You don't need to copy them here again. Just use the ones from drivers folder.

File: projects/ad9656_fmc/Makefile [hunk 11]
```diff
@@ -0,0 +1,6 @@
+TARGET := ad9656_fmc
```
Comment (mirceacaprioru) [inline projects/ad9656_fmc/Makefile:1]: you have a typo in the commit message: "recevied" -> "received"

File: projects/ad9656_fmc/src/app/ad9553.c [hunk 12]
```diff
@@ -0,0 +1,244 @@
+/***************************************************************************//**
+ * @file ad9553.c
+ * @brief Implementation of AD9553 Driver.
+ * @author DHotolea (dan.hotoleanu@analog.com)
```
Comment (mirceacaprioru) [inline projects/ad9656_fmc/src/app/ad9553.c:4]: It seems that you also added some new drivers with this PR. Separate the drivers in different commits and then just add a commit with the project.

File: projects/ad9656_fmc/src/app/ad9508.c (hunk 8 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9508.c:62]: Done in v7

File: projects/ad9656_fmc/src/app/ad9508.c (hunk 9 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9508.c:65]: Done in v7

File: projects/ad9656_fmc/src/app/ad9508.c (hunk 10 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9508.c:8]: These drivers are new, they are written by me.

File: projects/ad9656_fmc/Makefile (hunk 11 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/Makefile:1]: Changed in v7

**Review by dbogdan — APPROVED** [review]

**Comment by hotoleanudan** [comment]
Changelog v7:

- corrected commenting style and an extra line in ad9508.c.

File: projects/ad9656_fmc/src/app/ad9553.c (hunk 12 above)
Comment (hotoleanudan) [reply projects/ad9656_fmc/src/app/ad9553.c:4]: I discussed with Dragos Bogdan and we will leave it like it is until there will be another project that will require one of the two drivers. In that situation the setup functions will be modified to be more generic and the drivers will be included in the drivers folder.

Comment (mirceacaprioru) [reply projects/ad9656_fmc/src/app/ad9553.c:4]: Ack.

**Comment by mirceacaprioru** [comment]
Please do a rebase before merging this in order to fix the CI.

**Comment by hotoleanudan** [comment]
Changelog v8:

- done a rebase.

**Review by amiclaus — APPROVED** [review]
